pip/wheel.py

```diff
@@ -631,26 +657,126 @@ def supported(self, tags=None):
         return bool(set(tags).intersection(self.file_tags))
 
 
+class BuildEnvironment(object):
+    """Context manager to install build deps in a simple temporary environment
+    """
+    def __init__(self, no_clean=False):
+        self.prefix = tempfile.mkdtemp('pip-build-env-')
+        self.no_clean = no_clean
+
+    def __enter__(self):
+        self.save_path = os.environ.get('PATH', None)
+        self.save_pythonpath = os.environ.get('PYTHONPATH', None)
+
+        install_scheme = 'nt' if (os.name == 'nt') else 'posix_prefix'
+        install_dirs = get_paths(install_scheme, vars={
+            'base': self.prefix,
+            'platbase': self.prefix,
+        })
+
+        scripts = install_dirs['scripts']
+        if self.save_path:
+            os.environ['PATH'] = scripts + os.pathsep + self.save_path
+        else:
+            os.environ['PATH'] = scripts + os.pathsep + os.defpath
+
+        if install_dirs['purelib'] == install_dirs['platlib']:
+            lib_dirs = install_dirs['purelib']
+        else:
+            lib_dirs = install_dirs['purelib'] + os.pathsep + \
+                install_dirs['platlib']
+        if self.save_pythonpath:
+            os.environ['PYTHONPATH'] = lib_dirs + os.pathsep + \
+                self.save_pythonpath
+        else:
+            os.environ['PYTHONPATH'] = lib_dirs
+
+        return self.prefix
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.save_path is None:
+            os.environ.pop('PATH', None)
+        else:
+            os.environ['PATH'] = self.save_path
+
+        if self.save_pythonpath is None:
+            os.environ.pop('PYTHONPATH', None)
+        else:
+            os.environ['PYTHONPATH'] = self.save_pythonpath
+
+        if not self.no_clean:
+            rmtree(self.prefix)
+
+        return False  # Do not suppress exceptions
+
+
 class WheelBuilder(object):
     """Build wheels from a RequirementSet."""
 
     def __init__(self, requirement_set, finder, build_options=None,
-                 global_options=None):
+                 global_options=None, no_clean=False):
         self.requirement_set = requirement_set
         self.finder = finder
         self._cache_root = requirement_set._wheel_cache._cache_dir
         self._wheel_dir = requirement_set.wheel_download_dir
         self.build_options = build_options or []
         self.global_options = global_options or []
+        self.no_clean = no_clean
+
+    def _find_build_reqs(self, req):
+        """Get a list of the packages required to build the project, if any.
+
+        Build requirements can be specified in a pyproject.toml, as described
+        in PEP 518
+        """
+        if os.path.isfile(req.pyproject_toml):
+            with open(req.pyproject_toml) as f:
+                pp_toml = pytoml.load(f)
+            return pp_toml.get('build-system', {}).get('requires', [])
+
+        return []  # No pyproject.toml
```

Comment:
One other thing to consider is whether we should automatically add ``wheel`` if they list ``setuptools``,  since it will almost certainly be an error if they don't.